### ISSUE TITLE
Grpc random port (for testing)

### ISSUE THREAD

**Issue by tpodg** [quarkusio/quarkus#17465]
## Description

I am trying to write some tests which include grpc client and thought about spinning up an actual grpc server as part of my test. This way I could implement desired responses and avoid mocking. For this purpose it would be nice to have randomised port, something like already implemented for http server:
quarkus.http.port=0
or
quarkus.http.test-port=0

Would this make sense / be useful or would you suggest any better approach to this topic?

## Implementation ideas

(If you have any implementation ideas, they can go here, however please note that all design change proposals should be posted to [the Quarkus developer mailing list](mailto:quarkus-dev@googlegroups.com) (or [the corresponding Google Group](https://groups.google.com/group/quarkus-dev); see [the decisions process document](https://github.com/quarkusio/quarkus/blob/main/DECISIONS.adoc) for more information).

**Comment by michalszynkiewicz** [comment]
Why isn't the default port working for you in your scenario?

**Comment by tpodg** [comment]
It is in most of the cases, but currently I have a monorepo project with multiple applications. If the tests for multiple applications are running at the same time, there could be conflicts... Currently I have modified ports only for the tests using properties file, but is this enough to make sure everything will be fine on a build system or different local dev environments...?

**Comment by michalszynkiewicz** [comment]
got it, I think it makes sense. I don't know how easily doable is this. Are you interested in create a pull request for it?

**Comment by tpodg** [comment]
Sure, I can have a look into it. What would be preferred solution?
I guess testing the server shouldn't be a problem as "InProccess" types can be easily used, so this leaves us with tests on the client side where client and server ports should match.

My current workaround reads the client port via ConfigProvider and creates a server with the same port. If this is enough, existing property "quarkus.grpc.clients.'client'.port" could be used with some logic to randomise the port if it set to 0.

**Comment by michalszynkiewicz** [comment]
if you could ping me about it tomorrow, preferrably on Zulip, I will be more helpful :)

**Comment by ivansenic** [comment]
@tpodg Can you provide an example on how do you spin the the actual grpc server as the part of the test?

**Comment by Alexander-Krause-Glau** [comment]
@ivansenic Did you find a solution to your problem? I also want to start a gRPC server instance in my client's tests, but fail to do so.

**Comment by ivansenic** [comment]
> @ivansenic Did you find a solution to your problem? I also want to start a gRPC server instance in my client's tests, but fail to do so.

@Alexander-Krause-Glau I was able to have it setup with spinning a grpc server on my own for testing:

```
  @ConfigProperty(name = "quarkus.grpc.clients.my-service.host")
  String host;

  @ConfigProperty(name = "quarkus.grpc.clients.my-service.port")
  int port;

  @BeforeEach
  public void init() throws Exception {
    // init mock
    service = mock(..);
    interceptor = mock(ServerInterceptor.class);

    // mock interceptor
    lenient()
        .doAnswer(
            invocation -> {
              ServerCallHandler<?, ?> next = invocation.getArgument(2);
              return next.startCall(invocation.getArgument(0), invocation.getArgument(1));
            })
        .when(bridgeInterceptor)
        .interceptCall(any(), any(), any());

    // set up the server that runs on the target port
    SocketAddress address = new InetSocketAddress(host, post);
    server =
        NettyServerBuilder.forAddress(address)
            .directExecutor()
            .intercept(bridgeInterceptor)
            .addService(bridgeService)
            .build();
    server.start();
    channel = NettyChannelBuilder.forAddress(address).usePlaintext().build();
  }

  @AfterEach
  public void tearDown() throws Exception {
    channel.shutdown();
    server.shutdown();
    try {
      channel.awaitTermination(5, TimeUnit.SECONDS);
      server.awaitTermination(5, TimeUnit.SECONDS);
    } finally {
      channel.shutdownNow();
      server.shutdownNow();
    }
  }
```

**Comment by cescoffier** [comment]
This will be handled with the next gRPC server as it will use the same port as the http server.

**Comment by Alexander-Krause-Glau** [comment]
@ivansenic Thank you very much. I think that my test code lacked the mocked interceptor. I will try this in the next week due to holidays.

@cescoffier 

> This will be handled with the next gRPC server as it will use the same port as the http server.

I am not sure what you mean by "this".

**Comment by cescoffier** [comment]
I meant the ability to use a random port.

**Comment by Alexander-Krause-Glau** [comment]
@ivansenic Thank you again, I still had time left to test your approach. Can you also show how you inject the created `channel` in your client code? The class that I want to test uses the Mutiny-generated service code (that is annotated with `@GrpcClient`.

**Comment by ivansenic** [comment]
> @ivansenic Thank you again, I still had time left to test your approach. Can you also show how you inject the created `channel` in your client code? The class that I want to test uses the Mutiny-generated service code (that is annotated with `@GrpcClient`.

@Alexander-Krause-Glau Not sure if I understood you.. You don't need to do anything else except for what I wrote above.. So basically for me it's like this:

```
@ApplicationScoped
class SomeComponent {

   @GrpcClient("my-service") GrpcClient client;
}
```

and then in test

```
@QuarkusTest
class SomeComponentTest {

  @Inject SomeComponent cut;
  
  @ConfigProperty(name = "quarkus.grpc.clients.my-service.host")
  String host;

  @ConfigProperty(name = "quarkus.grpc.clients.my-service.port")
  int port;

  // plus before and after setup from  https://github.com/quarkusio/quarkus/issues/17465#issuecomment-1193825618
}
```

**Comment by sgrissom-rxss** [comment]
This seems to already work today on at least `2.11.2.Final`. Setting `quarkus.grpc.server.test-port=0` will setup a server on a random port, however unlike [VertxHttpRecorder](https://github.com/quarkusio/quarkus/blob/2.13.1.Final/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L1257-L1307), the `GrpcServerRecorder` does not update the system property to the actual port.

**Comment by cescoffier** [comment]
The new gRPC server uses the HTTP port (which can be random), so it fixes this issue.

**Comment by nahguam** [comment]
I don't believe this is fixed in `3.0.3.Final`. At least for the standard way of testing that the guide suggests by injecting a client with `@GrpcClient`.

Create a test project:

`quarkus create -x grpc`

Configure GRPC to run on the HTTP server and the make the HTTP test port random:

`application.properties`:
```
quarkus.grpc.server.use-separate-server=false
quarkus.http.test-port=0
```

The injected `@GrpcClient` in the test still tries to connect to the server on the default grpc test port of 9001. This can be worked around but as far as I can tell you'd have to manually create a channel/client by reading the `quarkus.http.test-port` system property that is dynamically set to the selected random port at test runtime. Is there a less intrusive way of getting the injected `@GrpcClient` to pick up the random test port?
